### PR TITLE
graph-struct-2: fixed problem with graph::Net templating

### PR DESCRIPTION
**Pull request by thequickf**
Went template woodoo way with graph::Net trait

```C++
class Node {
public:
    Node(int a, int b) : a(a), b(b) {}
    int a, b;
};

int main() {
    graph::Graph<int, graph::Net, graph::Weighted> g{{1, 2, 3}, {1, 3} /* <-- Setting source and sink here */};
    graph::Edge<Node> e{ {1, 2}, {1, 2} };
    g.addEdge({1, 2, 1  /* <-- weight */});
    g.addEdge({2, 3, 2});
    g.addEdge({3, 1, 3});
    return 0;
}
```

### CODE REVIEW

File: src/impl/meta_utils.hpp [hunk 1]
```diff
@@ -35,24 +44,59 @@ struct filter_impl<Condition, Head, Tail...> {
...
-namespace graph_impl {
+template <typename Node>
+struct replace_abstract_traits_impl<Node> {
+    using type = list<>;
+};
 
-template <template <typename> class Condition, typename... Ts>
-using filter = typename filter_impl<Condition, Ts...>::type;
+template <typename Node, typename Head, typename... Tail>
+struct replace_abstract_traits_impl<Node, Head, Tail...> {
```
Comment (isnullxbh) [inline src/impl/meta_utils.hpp:59]: This cannot be used as a customization point by the user.

File: src/impl/graph.hpp [hunk 2]
```diff
@@ -2,13 +2,22 @@
...
 #include <utility>
 
 namespace graph_impl {
 
+template<typename Node>
+class Net : public graph::GraphTrait {
+public:
+    Net(Node source, Node sink) :
+        source(std::move(source)), sink(std::move(sink)) {}
+    Node source, sink;
```
Comment (isnullxbh) [inline src/impl/graph.hpp:18]: Do we really want a copy, not a reference/pointer?

File: src/impl/meta_utils.hpp (hunk 1 above)
Comment (thequickf) [reply src/impl/meta_utils.hpp:59]: It shouldn't be.
User should inherit from `GraphTrait` and `EdgeTrait` for customization.

Comment (isnullxbh) [reply src/impl/meta_utils.hpp:59]: Ok, but how a user will do the same trick with its own trait (abstract -> impl)?

File: src/impl/graph.hpp (hunk 2 above)
Comment (thequickf) [reply src/impl/graph.hpp:18]: Yes. It makes more sense in my mind.
For example:
```C++
struct CustomNode {
    CustomNode(int id, int dp = 0) : id(id), dp(dp) {}
    int id;
    int dp;

    friend bool operator<(const CustomNode& lhn, const CustomNode& rhn) {
        return lhn.id < rhn.id;
    }
}
```
then
```C++
graph::Graph<CustomNode, graph::Net>g{{{1, 1}, {2, 1}, {3, 1}}, {1, 3}};
```
totally makes sense.

File: src/impl/meta_utils.hpp (hunk 1 above)
Comment (thequickf) [reply src/impl/meta_utils.hpp:59]: I guess user will use `CustomTrait<CustomNode>` and it should be ok.
I'll think about it.

File: src/impl/graph.hpp (hunk 2 above)
Comment (thequickf) [reply src/impl/graph.hpp:18]: Actually we kinda want reference/pointer but it should point on Node in container inside the Graph object.
I'll think about it.

**Review by isnullxbh — APPROVED** [review]

File: src/impl/graph.hpp [hunk 3]
```diff
@@ -30,9 +30,14 @@ class Graph<
     public GraphTraits... {
 public:
     Graph() = default;
-    Graph(std::initializer_list<Node> nodes,
+    Graph(std::initializer_list<Node> node_list,
           GraphTraits... graph_traits) :
-          nodes(std::move(nodes)), GraphTraits(std::move(graph_traits))... {};
+          nodes(std::move(node_list)), GraphTraits(std::move(graph_traits))... {
+        if (std::is_base_of<Net<Node>, typeof(*this)>::value) {
```
Comment (isnullxbh) [inline src/impl/graph.hpp:36]: There is no operator `typeof` in C++

**Pull request merged**